File: src/constraints/no_constraints.rs
use super::Constraint;

/// The whole space, no constraints
#[derive(Default, Clone, Copy)]
pub struct NoConstraints {}

impl NoConstraints {
    /// Constructs new instance of `NoConstraints`
    ///
    #[must_use]
    pub fn new() -> NoConstraints {
        NoConstraints {}
    }
}

impl<T> Constraint<T> for NoConstraints {
    fn project(&self, _x: &mut [T]) {}

    fn is_convex(&self) -> bool {
        true
    }
}

File: open-codegen/opengen/constraints/no_constraints.py
from .constraint import Constraint


class NoConstraints(Constraint):
    """
    This is the entire :math:`{\\rm I\!R}^n`

    """

    def __init__(self):
        pass

    def distance_squared(self, u):
        return 0.0

    def project(self, u):
        return u

    def is_convex(self):
        return True

    def is_compact(self):
        return False
